s3fs/core.py

```diff
@@ -1716,11 +1740,13 @@ async def _rm(self, path, recursive=False, **kwargs):
         files = [p for p in paths if self.split_path(p)[1]]
         dirs = [p for p in paths if not self.split_path(p)[1]]
         # TODO: fails if more than one bucket in list
-        await asyncio.gather(
-            *[
+        await _run_coros_in_chunks(
+            [
                 self._bulk_delete(files[i : i + 1000])
                 for i in range(0, len(files), 1000)
-            ]
+            ],
+            batch_size=3,
```

Comment:
Where does this batch size come from? I would have thought it could be a much bigger number.

Reply:
With this value, how long did it take to delete ~1000 keys or more?

Reply:
This is in the context of many dask tasks doing the same in parallel?

Reply:
> This is in the context of many dask tasks doing the same in parallel?

No, this is in the context of a bulk delete from a single `S3FileSystem` instance. That occurs in a dask `Client` when `overwrite=True`, but can reproduced by itself as per https://github.com/fsspec/s3fs/issues/484#issuecomment-1113723501

Reply:
> these must come back in about 1ms, right?

I think they come back in ~1s, but are batched into bulk delete calls of 1000 keys.

> Now we have the tricky problem that the "best" number will depend on where this is called from.

Yeah, it's tough to come up with a "best" number. That's what I was attempting to do in #619.

Reply:
Of course, so sorry - these are thousands of deletes! Please ignore my hesitation in that case. It totally makes sense to send ~3000 requests and wait for them all to end before sending more.

Reply:
I just re-ran my testing scripts with some timing information -- in my case each 1000-key bulk delete took 1-2 seconds, though I don't know how reproducible that is in other contexts.

Reply:
+1